Zarr: Allow int64 attributes

eschnett opened the pull request:
## What does this PR do?

The Zarr file format stores attributes in JSON. This PR stores int64 values as integer JSON values instead of converting them to floating-point values.

## What are related issues/pull requests?

https://github.com/OSGeo/gdal/issues/10062

## Tasklist

 - [x] Make sure code is correctly formatted (cf [pre-commit configuration](https://gdal.org/development/dev_practices.html#commit-hooks))
 - [x] Add test case(s)
 - [x] Add documentation
 - [x] Updated Python API documentation (swig/include/python/docs/)
 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

rouault commented on gcore/gdal.h:
```diff
@@ -2513,18 +2517,27 @@ void CPL_DLL GDALAttributeFreeRawResult(GDALAttributeH hAttr, GByte *raw,
                                         size_t nSize);
 const char CPL_DLL *GDALAttributeReadAsString(GDALAttributeH hAttr);
 int CPL_DLL GDALAttributeReadAsInt(GDALAttributeH hAttr);
+GInt64 CPL_DLL GDALAttributeReadAsLong(GDALAttributeH hAttr);
```
Please avoid the "Long" name as it is never clear how large that type is. Use "Int64" instead. Applies to all new additions. You can also use int64_t . GInt64 is legacy at that point

rouault commented:
You can setup pre-commit with https://gdal.org/development/dev_practices.html#commit-hooks to have automatic code formatting with the version of clang we've standardized on

eschnett commented:
Thanks for the pointers so far.

eschnett commented:
I have a few comments and questions:
- The class `GDALAttribute` defines many `Write` functions, and one function `WriteInt` with a different name. I have added a function `WriteInt64` there.
- I ended up adding functions to the core of gdal instead of just modifying the zarr driver as I originally anticipated. I added functions `WriteInt64`, `WriteInt64Array`, and respective reader functions as well.
- This change lets me add `Int64` values to Zarr files. However, it doesn't handle `UInt64` values yet. Should I add support for this as well? This would add a few more functions to the core of gdal. Currently, `UInt64` are converted to `Double`, similar to the way in which `Int64` values were handled before.
- I didn't find any explicitly written documentation for `WriteInt` or `WriteDouble`, and thus there wasn't any place for me to document the functions I added. I assume it's all auto-generated?

rouault commented:
Excellent work! I didn't thought you would have to dwelve into SWIG stuff...

> * The class `GDALAttribute` defines many `Write` functions, and one function `WriteInt` with a different name

I don't recall the exact reason why we didn't use C++ name overloading here. I'm not sure there's a good reason, especially since the C++ method seems only to be used in the code base by the GDALAttributeWriteInt() C function.

> I have added a function WriteInt64 there. 

sounds good

> This change lets me add Int64 values to Zarr files. However, it doesn't handle UInt64 values yet. Should I add support for this as well?

could be done as a follow up if needed

> I didn't find any explicitly written documentation for WriteInt or WriteDouble

Not sure to understand. They have Doxygen comments, and they pop up in https://gdal.org/api/gdalattribute_cpp.html#_CPPv4N13GDALAttribute8WriteIntEi

Your changes caused a failure in one of the netcdf_multidimensional.py test, but I finally figured this was actually a long standing bug in the netCDF driver. I've added a fix in your branch for that.

I've also removed a bit of your added typemaps as some of them were actually unneeded
